concurrency-primer.tex

```diff
@@ -747,6 +747,178 @@ \section{Do we always need sequentially consistent operations?}
+
+Although sequential consistency is considered the ``golden standard'' for multi-threaded programs, its many constraints limit performance optimization.
+As a result, it is rarely implemented in modern processors.
+Instead, more relaxed memory models are used, such as the total store order (TSO) memory model adopted by the x86 architecture.
+One can envision the hardware roughly as follows:
+
+\begin{center}
+\includegraphics[keepaspectratio,width=0.7\linewidth]{images/hw-tso}
+\captionof{figure}{The memory model of x86-TSO hardware.}
+\label{hw-tso}
+\end{center}
+
+All processors can read from a single shared memory, but each processor writes only to its own write queue.
+
+Consider the following Write Queue (Store Buffer) Litmus Test:
+
+\begin{ccode}
+// Litmus Test: Write Queue (Store Buffer)
+int x = 0;
+int y = 0;
+
+// Thread 1        // Thread 2
+x = 1;             y = 1;
+r1 = y;            r2 = x;
+\end{ccode}
+
+Sequential consistency does not allow \monobox{r1 = r2 = 0}, but TSO does.
+In a sequentially consistent memory model, \monobox{x = 1} or \monobox{y = 1} must be written first, followed by the read operations, so \monobox{r1 = r2 = 0} cannot occur.
+However, under the TSO memory model, the write operations from both threads might still be in their respective queues when the read operations occur, allowing \monobox{r1 = r2 = 0}.
+
+Non-sequentially consistent hardware typically supports additional memory barrier (fence) instructions to control the order of read and write operations.
+These barriers ensure that writes before the barrier are completed (queues emptied) before any subsequent reads are performed.
+
+\begin{ccode}
+// Thread 1           // Thread 2
+x = 1;                y = 1;
+barrier;              barrier;
+r1 = y;               r2 = x;
+\end{ccode}
+
+The reason total store order (TSO) is named as such is because once a write operation reaches shared memory, it indicates that all processors are aware that the value has been written.
+There will be no situation where different processors see different values.
+That is, the following litmus test will not have \monobox{r1 = 1}, \monobox{r2 = 0}, \monobox{r3 = 1}, but \monobox{r4 = 0}.
+
+Consider the following Independent Reads of Independent Writes (IRIW) Litmus Test:
+
+\begin{ccode}
+// Litmus Test: Independent Reads of Independent Writes (IRIW)
+int x = 0;
+int y = 0;
+
+// Thread 1    // Thread 2    // Thread 3    // Thread 4
+x = 1;         y = 1;         r1 = x;        r3 = y;
+                              r2 = y;        r4 = x;
+\end{ccode}
+
+Once Thread 3 reads \monobox{r1 = 1}, \monobox{r2 = 0}, it indicates that the write \monobox{x = 1} reached shared memory before \monobox{y = 1}.
+If at this point Thread 4 reads \monobox{r3 = 1}, it means both writes \monobox{y = 1} and \monobox{x = 1} are visible to Thread 4, so \monobox{r4} can only be \monobox{1}.
+We can say "Thread 1's write to \monobox{x}" happens before "Thread 2's write to \monobox{y}".
```

Comment:
Be aware of the quotation marks.
See https://www.maths.tcd.ie/~dwilkins/LaTeXPrimer/QuotDash.html